File: src/engine/ChangeLens.Engine/Repositories/Models/RepositoryRestoreResult.cs
using System.Text.Json.Serialization;

namespace ChangeLens.Engine.Repositories.Models;

/// <summary>
///     Represents the tagged startup repository-restoration result.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "state")]
[JsonDerivedType(typeof(NoRepositoryRestoreResult), "none")]
[JsonDerivedType(typeof(RestoredRepositoryResult), "restored")]
internal abstract record RepositoryRestoreResult;

File: src/desktop/ui/src/Repositories/Models/RepositoryRestoreResult.ts
import type { RepositoryDescriptor } from "./RepositoryDescriptor";

export type RepositoryRestoreResult =
  | { readonly state: "none" }
  | {
      readonly state: "restored";
      readonly repositoryId: string;
      readonly repository: RepositoryDescriptor;
      readonly preferredTarget: string | null;
    };
